src/expr/attribute_internals.h

```diff
@@ -145,16 +149,372 @@ inline uint64_t GetBitSet(uint64_t bit)
+  {
+    // reach inside the iterator to get L1/L2 positions
+    auto nextL2It = it.d_l1It->second.erase(it.d_l2It);
+
+    iterator nextIt(this, it.d_l1It, nextL2It);
+    nextIt.legalize();
+
+    if (it.d_l1It->second.empty())
+    {
+      // this erase has made the L2 map empty. delete it:
+      d_storage.erase(it.d_l1It);
+    }
+
+    return nextIt;
+  }
+
+  template <typename Iter>
+  void insert(Iter beg, Iter end)
+  {
+    using Entry = typename std::iterator_traits<Iter>::value_type;
+    using Entries = std::vector<Entry>;
+    using EntryIt = typename Entries::iterator;
+    Entries entries(beg, end);
+
+    // sort by second (NodeValue*) then first (uint64_t)
+    std::sort(
+        entries.begin(), entries.end(), [](const Entry& a, const Entry& b) {
+          return (a.first.second < b.first.second)
+                 || ((a.first.second == b.first.second)
+                     && (a.first.first < b.first.first));
+        });
+
+    auto find_different_nv = [](NodeValue* nv, EntryIt first, EntryIt last) {
+      return std::find_if(
+          first, last, [nv](const Entry& a) { return a.first.second != nv; });
+    };
+
+    std::size_t l1_unique_count = 0;
+    auto last = entries.end();
+    for (EntryIt it = entries.begin(); it != last;
+         it = find_different_nv(it->first.second, it, entries.end()))
+    {
+      ++l1_unique_count;
+    }
+    d_storage.reserve(d_storage.size() + l1_unique_count);
+
+    for (EntryIt it = entries.begin(); it != last;)
+    {
+      EntryIt chunk_end = std::find_if(it, entries.end(), [it](const Entry& v) {
+        return v.first.second != it->first.second;
+      });
+      // add to l2 map
+      auto& l2 = d_storage[it->first.second];
+      l2.reserve(l2.size() + std::distance(it, chunk_end));
+      for (; it != chunk_end; ++it)
+      {
+        l2.emplace(it->first.first, it->second);
+      }
+    }
+  }
+
+  void swap(AttrHash& other) { std::swap(d_storage, other.d_storage); }
+
+  V& operator[](std::pair<uint64_t, NodeValue*> p)
+  {
+    return d_storage[p.second][p.first];
+  }
+
+  void clear() { d_storage.clear(); }
+
+  const_iterator find(std::pair<uint64_t, NodeValue*> p) const
+  {
+    typename Storage::const_iterator it1 = d_storage.find(p.second);
+    if (it1 == d_storage.end()) return const_iterator();
+
+    typename IdMap::const_iterator it2 = it1->second.find(p.first);
+    if (it2 == it1->second.end()) return const_iterator();
+
+    return const_iterator(this, it1, it2);
+  }
+
+  iterator find(std::pair<uint64_t, NodeValue*> p)
+  {
+    typename Storage::iterator it1 = d_storage.find(p.second);
+    if (it1 == d_storage.end()) return iterator();
+
+    typename IdMap::iterator it2 = it1->second.find(p.first);
+    if (it2 == it1->second.end()) return iterator();
+
+    return iterator(this, it1, it2);
+  }
+
+  void eraseBy(NodeValue* nv) { d_storage.erase(nv); }
+
+ private:
+  Storage d_storage;
+};
+// };/* class AttrHash<> */
```

Comment:
```suggestion
```

Reply:
I think I'll keep just the `/* class AttrHash<> */` bit, if that's OK with you, since it's consistent with other classes.

Reply:
Yes that's fine!